File: test/2022-robot-command/src/main/cpp/commands/DriveCartesian.cpp
#include "commands/DriveCartesian.h"

#include <utility>

DriveCartesian::DriveCartesian(Drivetrain* subsystem, 
	std::function<double()> forward, std::function<double()> right, std::function<double()> rotation, 
	std::function<bool()> brakeButton, std::function<bool()> coastButton, std::function<int()> maxSpeedPov) 
	: drivetrain(subsystem), x(std::move(forward)), y(std::move(right)), z(std::move(rotation)),
	doBrake(std::move(brakeButton)), doCoast(std::move(coastButton)), maxSpeedPOV(std::move(maxSpeedPov))
{
	AddRequirements(subsystem);
}

void DriveCartesian::Execute() {
	xSquare = x() * fabs(x()); //multiply each axis by its absolute value, this makes acceleration exponential rather than linear
	ySquare = y() * fabs(y());
	zSquare = z() * fabs(z());

	if (fabs(x()) > DrivetrainConst::JOYSTICKDEADZONE || fabs(y()) > DrivetrainConst::JOYSTICKDEADZONE || fabs(z()) > DrivetrainConst::JOYSTICKDEADZONE)
		drivetrain->Drive(
			CloserToZero(xSquare, xAccelerationCap.Calculate(xSquare)),
			CloserToZero(ySquare, yAccelerationCap.Calculate(ySquare)), 
			CloserToZero(zSquare, zAccelerationCap.Calculate(zSquare)));
	else
		drivetrain->Drive(0, 0, 0);

	UpdateBrake();

	//max speed:
	if (maxSpeedPOV() == 0 && maxSpeedPOVPrevState != 0) {
		drivetrain->AddToMaxSpeed(0.1f);
	}
	if (maxSpeedPOV() == 180 && maxSpeedPOVPrevState != 180) {
		drivetrain->AddToMaxSpeed(-0.1f);
	}
	maxSpeedPOVPrevState = maxSpeedPOV();
}

void DriveCartesian::End(bool interrupted) {
	drivetrain->Drive(0, 0, 0);
}

void DriveCartesian::UpdateBrake() {
	if (doBrake() && !doBrakePrevState) {
		drivetrain->ActivateBreakMode(true);
		doBrakePrevState = true;
	}
	else if (!doBrake() && doBrakePrevState)
		doBrakePrevState = false;
	else if (doCoast() && !doCoastPrevState) {
		drivetrain->ActivateBreakMode(false);
		doCoastPrevState = true;
	}
	else if (!doCoast() && doCoastPrevState) {
		doCoastPrevState = false;
	}
}

double DriveCartesian::CloserToZero(double value1, double value2) {
	if (fabs(value1) < fabs(value2)) {
		return value1;
	}
	else {
		return value2;
	}
}

File: test/event-bot/src/main/java/frc/robot/commands/DriveCartesian.java
package frc.robot.commands;

import java.util.function.BooleanSupplier;
import java.util.function.DoubleSupplier;

import edu.wpi.first.math.filter.SlewRateLimiter;
import edu.wpi.first.wpilibj2.command.CommandBase;
import frc.robot.Constants.DrivetrainConst;
import frc.robot.subsystems.Drivetrain;

public class DriveCartesian extends CommandBase {
    Drivetrain drivetrain;

    DoubleSupplier x;
    DoubleSupplier y;
    DoubleSupplier z;

    double xSquare;
    double ySquare;
    double zSquare;

    SlewRateLimiter xAccelerationCap = new SlewRateLimiter(3);
    SlewRateLimiter yAccelerationCap = new SlewRateLimiter(3);
    SlewRateLimiter zAccelerationCap = new SlewRateLimiter(3);

    BooleanSupplier povLeft; //whether the POV is pressed towards the left

    public DriveCartesian(Drivetrain drivetrain_p, 
        DoubleSupplier forward, DoubleSupplier right, DoubleSupplier rotation,
        BooleanSupplier povLeft_p)
    {
        drivetrain = drivetrain_p;

        x = forward;
        y = right;
        z = rotation;

        povLeft = povLeft_p;

        addRequirements(drivetrain_p);
    }

    @Override
    public void execute() {
        xSquare = x.getAsDouble() * Math.abs(x.getAsDouble());
        ySquare = y.getAsDouble() * Math.abs(y.getAsDouble());
        zSquare = z.getAsDouble() * Math.abs(z.getAsDouble());

        if (Math.abs(x.getAsDouble()) > DrivetrainConst.JoystickDeadzone || Math.abs(y.getAsDouble()) > DrivetrainConst.JoystickDeadzone || Math.abs(z.getAsDouble()) > DrivetrainConst.JoystickDeadzone) {
            drivetrain.drive(closerToZero(xSquare, xAccelerationCap.calculate(xSquare)), 
                closerToZero(ySquare, yAccelerationCap.calculate(ySquare)), 
                closerToZero(zSquare, zAccelerationCap.calculate(zSquare)));
        }
        else
            drivetrain.drive(0, 0, 0);
    }

    @Override
    public void end(boolean interrupted) {
        drivetrain.drive(0, 0, 0);
    }

    private double closerToZero(double value1, double value2) {
        if (Math.min(Math.abs(value1), Math.abs(value2)) == Math.abs(value1))
            return value1;
        else
            return value2;
    }
}
